### ISSUE TITLE
Mypy not correctly following logic with the bitwise xor operator.

### ISSUE THREAD

**Issue by BSchilperoort** [python/mypy#13379]
**Bug Report**
Mypy does not correctly follow logic with the bitwise xor operator, raising an error where it should not.

**To Reproduce**

Minimum viable example:
```py
from typing import Optional
from typing import List


def append_lists(
    list1: Optional[List[float]] = None,
    list2: Optional[List[float]] = None,
    ):

    if (list1 is None) and (list2 is None):
        list1 = []
        list2 = []
    elif (list1 is None) ^ (list2 is None):
        raise ValueError

    list1.append(0)
    list2.append(0)

```
Raises:
```
set_title.py:16: error: Item "None" of "Optional[List[float]]" has no attribute "append"
set_title.py:17: error: Item "None" of "Optional[List[float]]" has no attribute "append"
```

**Expected Behavior**
No error should be raised, as both `list1` and `list2` will both by of type `List` at line 16 and 17, otherwise a `ValueError` will have been raised.

| list1  | list2  | Outcome                                       |
|--------|--------|-----------------------------------------------|
| `None` | `None` | Both `None`: set as empty list, append with 0 |
| `None` | `[]`   | `raises ValueError`                           |
| `[]`   | `[]`   | Both valid: appends 0                         |
| `[]`   | `None` | `raises ValueError`                           |

This code is functionally the same (and is probably more clear than the `^` operator), but will not raise an error:
```py
def append_lists(
    list1: Optional[List[float]] = None,
    list2: Optional[List[float]] = None,
    ):

    if (list1 is None) and (list2 is None):
        list1 = []
        list2 = []
    elif (list1 is None) or (list2 is None):
        raise ValueError

    list1.append(0)
    list2.append(0)

```

**Your Environment**
- Mypy version used: 0.971
- Mypy command-line flags: None
- Mypy configuration options from `mypy.ini` (and other config files): None
- Python version used: 3.10.5
- Operating system and version: Windows 10

**Comment by AlexWaygood** [comment]
Mypy (and all other type checkers that I know of) has to hard-code support for each method of type-narrowing it supports. It doesn't have a generalised understanding of all possible Python constructs and whether or not each one narrows types in a safe way — that would require mypy to be far more complex than it already is.

You can find a list of supported type-narrowing constructs here: https://mypy.readthedocs.io/en/latest/type_narrowing.html

I've never seen anybody use the bitwise xor operator for type narrowing before, so I think it's really unlikely this is going to be supported by mypy in the foreseeable future. As such, I'm closing this issue.
